kart/parse_args.py

```diff
@@ -75,29 +61,89 @@ def get_arg_type(repo, arg, allow_options=True, allow_commits=True, allow_filter
-            raise click.UsageError(f"No such option: {arg}")
+            options.append(arg)
+    return options, others
+
+
+def _append_kwargs_to_options(options, kwargs, allow_options):
+    if not kwargs:
+        return
+    assert allow_options
+
+    for option_name, option_val in kwargs.items():
+        option_name = option_name.replace("_", "-", 1)
+        t = type(option_val)
+        if t == int or t == str:
+            options.append(f"--{option_name}={option_val}")
+        elif t == tuple:
+            options.extend([f"--{option_name}={o}" for o in option_val]),
+        elif t == bool and option_val:
+            options.append(f"--{option_name}")
 
-    if allow_commits:
-        if arg == "[EMPTY]" or ".." in arg:
-            return ArgType.COMMIT
 
-        try:
-            repo.resolve_refish(arg)
-            return ArgType.COMMIT
-        except (KeyError, ValueError, pygit2.InvalidSpecError):
-            pass
+HEAD_PATTERN = re.compile(r"^HEAD\b", re.IGNORECASE)
 
-    if allow_filters:
-        return ArgType.FILTER
+HINT = "Use '--' to separate paths from revisions, like this:\n'kart <command> [<revision>...] -- [<filter>...]'"
+SILENCING_HINT = "To silence this warning, use '--' to separate paths from revisions, like this:\n'kart <command> [<revision>...] -- [<filter>...]'\n"
 
-    raise click.UsageError(
-        f"Argument not recognised as a valid commit, ref, or range: {arg}"
-    )
 
+def _is_revision(repo, arg, warning_set):
+    # These things *could* be a path, but in that case the user should add a `--` before this arg to
+    # disambiguate, and they haven't done that here.
+    if arg == "[EMPTY]" or ".." in arg or HEAD_PATTERN.match(arg) or arg.endswith("^?"):
+        return True
 
-def parse_commits_and_filters(
+    filter_path = arg.split(":", maxsplit=1)[0]
+    head_tree = repo.head_tree
+    is_useful_filter_at_head = head_tree and filter_path in head_tree
+
+    if ":" in arg and not is_useful_filter_at_head:
+        click.echo(
+            f"Assuming '{arg}' is a filter argument (that doesn't match anything at HEAD)",
+            err=True,
+        )
+        warning_set.add(SILENCING_HINT)
+        return False
+
+    try:
+        repo.resolve_refish(arg)
+        is_revision = True
+    except (KeyError, ValueError, pygit2.InvalidSpecError):
+        is_revision = False
+
+    if is_revision and not is_useful_filter_at_head:
+        return True
+    elif is_useful_filter_at_head and not is_revision:
+        return False
+    elif is_revision and is_useful_filter_at_head:
+        raise click.UsageError(
+            f"Ambiguous argument '{arg}' - could be either a revision or a filter\n{HINT}"
+        )
+    else:
+        raise NotFound(
+            f"Ambiguous argument '{arg}' - doesn't appear to be either a revision or a filter\n{HINT}"
+        )
+
+
+def _disambiguate_revisions_and_filters(repo, args):
+    revisions = []
+    filters = []
+    warning_set = set()
```

Comment:
not sure this needs to be a set; it looks like you're only adding SILENCING_HINT, so there should be other ways to dedupe that. Making it a set means the output will be nondeterministically ordered if you ever add a second thing to the set.

Perhaps we should configure the [warnings](https://docs.python.org/3/library/warnings.html) module? It will de-dupe warnings automatically if configured correctly

Reply:
... the warnings module looks great for deprecating things in modules that will be imported by third parties, etc etc. But I am just trying to print something once: I don't need auto line numbers + module names, or configuration options, and I especially don't want it to remember which warnings it has already printed for any longer than this function call (as could happen with our testing framework / kart helper). It's not beautiful how it is but I've just spent 10 minutes thinking about this and I think it's not really worth spending any longer on it - there's actually nothing measurably wrong with it right now

Reply:
> as could happen with our testing framework

AFAIK pytest resets the warnings state between each test, by default. It also has really good support for the warnings module if you want to check for warnings in your tests: https://docs.pytest.org/en/latest/how-to/capture-warnings.html

> kart helper

yep, that could use some work. Looks like there's no simple way to reset the warning registry globally: https://bugs.python.org/issue21724 - which we'd need to make this work well for kart helper

So in that case, I guess leave this as it is 👍